### PR TITLE
SUBSTRING_INDEX() func

### LINKED ISSUE #3033: SUBSTRING_INDEX() | Return a substring from a string before the specified number of occurrences of the delimiter

### PR DESCRIPTION
**Pull request by mshauneu**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Impl of SUBSTRING_INDEX() func

## Changelog

- New Feature

## Related Issues

Fixes #3033 

## Test Plan

Stateless Tests

### CODE REVIEW

File: website/databend/docs/sqlstatement/string-functions/substring_index.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+---
```
Comment (BohuTANG) [inline website/databend/docs/sqlstatement/string-functions/substring_index.md:1]: +1

File: common/functions/src/scalars/strings/substring_index.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        }
+        Ok(DataType::String)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
+        let s_column = columns[0].column().cast_with_type(&DataType::String)?;
+        let d_column = columns[1].column().cast_with_type(&DataType::String)?;
+        let c_column = columns[2].column().cast_with_type(&DataType::Int64)?;
+
+        let r_column: DataColumn = match (s_column, d_column, c_column) {
```
Comment (sundy-li) [inline common/functions/src/scalars/strings/substring_index.rs:84]: That's great.

But I have some thoughts, how about forcing the third argument to be Constant?
I did not find any reason to use dynamic count.

Thus we can reduce lots of branches.

**Review by sundy-li — APPROVED** [review]

File: common/functions/src/scalars/strings/substring_index.rs (hunk 2 above)
Comment (sundy-li) [reply common/functions/src/scalars/strings/substring_index.rs:84]: From ClickHouse doc:

https://github.com/ClickHouse/ClickHouse/blob/cf0fbbbc1704e8d232039547005424830a0c82d9/docs/en/sql-reference/functions/string-functions.md

**Pull request merged**